### ISSUE TITLE
C binding for libdar?

### ISSUE THREAD

**Issue by piperun** [Edrusb/DAR#55]
I've been eying this project for some time and been wanting to use it in Rust/Dart, unfortunately there's no C binding to more easily dig your teeth into.
I noticed there is a python binding to my surprise but not C.

**Comment by Edrusb** [comment]
correct, libdar is written in C++ and makes a big use of classes and objects. Though, one could argue that using an object is almost equivalent as adding a first argument (a pointer to the data-structure of the object, which can be an opaque one from API user point of view) to any method the class provides be converted to a C function... 

If you are willing to set this up, let me know, I can guide you from the starting point (which code/branch to use from git)  which part to focus on, up to what automatic/systematic naming could worth considering... Unfortunately, I have not free/personal time left for to do that alone, this is why I cannot add it to my todo list.

**Comment by piperun** [comment]
Thank you for the reply!
I will admit I'm not very well versed enough in C to form a proper C bindings, as I plan on investing time in learning rust in the near future that might be an option instead?

**Comment by Edrusb** [comment]
Searching Internet, it seems complicated to call C++ routines from rust due to C++ name mangling... it is advised to pass through C, which is probably the reason why you were looking for C API to libdar...
 
The path seems less complicated to reach python code from Rust, though this may not be totally straight forward... but libdar provides a python API as you know...

my 2 cents...

**Comment by piperun** [comment]
That is fair enough, I did find 2 projects that seem to have C bindings for libdar, but they are very old/incomplete:
https://sourceforge.net/projects/darbinding/
And this is the binding DarGUI uses:
https://sourceforge.net/p/dargui/code/HEAD/tree/libdargui/

**Comment by Edrusb** [comment]
I'm sorry, I'd like to have more time to work on dar/libdar but I still have to pay my bills... and dar/libdar will stay free. So there are priorities and choices to take to match my available free time. C binding is not very demanded and while it does not seems very risky nor very complex to address (well, seen from far away...) it requires an important investment in time... something I have not for now... being busy some free software project...

**Comment by piperun** [comment]
I'm not sure I follow, I only stated I found these.
It was more of a look what I found rather than some demand, do apologize if it came out as such.

**Comment by Edrusb** [comment]
This was unclear on my side what to do with these projects. Thanks for this feedback and no worries at all!

**Comment by Edrusb** [comment]
C API should not be a big challenge but requires a lot of time to setup:
the principle is for each C++ class surfacing in the API is the same

assuming class foo is defined 
as such:

```
class foo [:public parent class]
{
  public:
       foo(const std::string & arg); // this is the class constructor
       
       std::string get_val() const         // method 1
       void set_val(const std::string); // method 2

      /* ignore the "protected:" and "private:" part if any, they are not part of the API */
}
```

the C binding could define: a function for the constructor 
another for the (here implicit) destructor
yet another for each method

```
void* create_foo(char *arg)
{
       void* ret = new (std::nothrow) foo(std::string(arg));
       return ret; /* returning NULL if allocation failed */
}

void delete_foo(void *ptr)
{
       foo* obj = (foo*)(ptr);
       delete obj;
}

int foo_get_val(void *obj, char* ret, unsigned int ret_size)
{
       foo* obj = (foo*)(ptr);
       std::string val;

       if(foo == nullptr)
          return -1;

      val = foo->get_val();
  
      strncpy(ret, val.c_str(), ret_size);
      ret[ret_size - 1] = '\0';

      return 0;
}

void foo_set_val(void *obj, char* ret, unsigned int ret_size)
{
       foo* obj = (foo*)(ptr);
 
       if(foo == nullptr)
          return -1;

       ret[ret_size - 1] = '\0';
       foo->set_val(std::string(ret));
}
 
```
This is systematic, for each class of the API, and their public methods but there is also to translate exception to exit code and or messages encapsulated in exception, this could be a (void*) pointer to return instead of an unsigned int as here, having NULL (0) meaning no error, else a pointer to an Egeneric exception object and a function to get the code/message from that object and onèfly deleting it.

It does not seem to have a big complexity at this distance, if someone which to take this as an exercise let me know!
